File: ios/Classes/BraintreeCheckoutFlutterPlugin.swift
import Flutter
import UIKit
import Braintree

public class BraintreeCheckoutFlutterPlugin: NSObject, FlutterPlugin {
    private var flutterResult: FlutterResult?
    var universalLinkURL: URL?
    
    public static var shared = BraintreeCheckoutFlutterPlugin()
    
    override private init() {}

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "braintree_checkout_flutter", binaryMessenger: registrar.messenger())
        let instance = BraintreeCheckoutFlutterPlugin.shared
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        self.flutterResult = result

        switch call.method {
        case Constants.DATA_INFO_METHOD_KEY:
            if let args = call.arguments as? [String: Any] {
                collectDeviceData(arguments: args, result: result)
            } else {
                result(FlutterError(code: Constants.ERROR_KEY, message: "Invalid arguments for getData", details: nil))
            }

        case Constants.PAYPAL_PAYMENT_METHOD_KEY:
            PayPalHandler.handle(arguments: call.arguments, result: result)

        case Constants.VENMO_PAYMENT_METHOD_KEY:
            VenmoHandler.handle(arguments: call.arguments, result: result)
            
        case Constants.IS_VENMO_APP_INSTALLED:
            VenmoHandler.isVenmoAppInstalled(result: result)
            
        case Constants.TOKENIZE_CARD_METHOD_KEY:
            tokenizeCard(arguments: call.arguments, result: result)
        
        case Constants.THREE_D_SECURE_METHOD_KEY:
            ThreeDSecureHandler.handle(arguments: call.arguments, result: result)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func isBraintreeUniversalLink(url: URL) -> Bool {
        guard let universalLinkURL else {
            return false
        }

        return url.scheme == universalLinkURL.scheme &&
               url.host == universalLinkURL.host &&
               url.path.hasPrefix(universalLinkURL.path)
    }

    private func collectDeviceData(arguments: [String: Any], result: @escaping FlutterResult) {
        guard let token = arguments[Constants.TOKEN_KEY] as? String else {
            flutterResult?(FlutterError(code: Constants.ERROR_KEY, message: "Token is required for data collection", details: nil))
            return
        }

        guard let braintreeClient = BTAPIClient(authorization: token) else {
            flutterResult?(FlutterError(code: Constants.ERROR_KEY, message: "Invalid Braintree client", details: nil))
            return
        }

        let dataCollector = BTDataCollector(apiClient: braintreeClient)
        dataCollector.collectDeviceData { (deviceData, error) in
            if let error = error {
                result(FlutterError(code: Constants.ERROR_KEY, message: error.localizedDescription, details: nil))
            } else if let deviceData = deviceData {
                result(deviceData)
            } else {
                result(FlutterError(code: Constants.ERROR_KEY, message: "Device data collection failed", details: "No data and no error returned"))
            }
        }
    }
    
    private func tokenizeCard(arguments: Any?, result: @escaping FlutterResult) {
        guard let args = arguments as? [String: Any] else { return }
        
        guard
            let token = args[Constants.TOKEN_KEY] as? String,
            let cardholderName = args[Constants.CARDHOLDER_NAME_KEY] as? String,
            let cardNumber = args[Constants.CARD_NUMBER_KEY] as? String,
            let expirationMonth = args[Constants.EXPIRATION_MONTH_KEY] as? String,
            let expirationYear = args[Constants.EXPIRATION_YEAR_KEY] as? String,
            let cvv = args[Constants.CVV_KEY] as? String
        else {
            flutterResult?(FlutterError(code: Constants.ERROR_KEY, message: "Token, card number, expiration month, expiration year and cvv are required for tokenization", details: nil))
            return
        }
        
        guard let braintreeClient = BTAPIClient(authorization: token) else {
            flutterResult?(FlutterError(code: Constants.ERROR_KEY, message: "Invalid Braintree client", details: nil))
            return
        }
        
        let cardClient = BTCardClient(apiClient: braintreeClient)
        let card = BTCard()
        card.cardholderName = cardholderName
        card.number = cardNumber
        card.expirationMonth = expirationMonth
        card.expirationYear = expirationYear
        card.cvv = cvv
        cardClient.tokenize(card) { tokenizedCard, error in
            if let error = error {
                result(FlutterError(code: Constants.ERROR_KEY, message: error.localizedDescription, details: nil))
            } else if let tokenizedCard = tokenizedCard {
                let json: [String: Any?] = [
                    "nonce": tokenizedCard.nonce,
                    "isDefault": tokenizedCard.isDefault,
                    "cardType": tokenizedCard.type,
                    "lastTwo": tokenizedCard.lastTwo,
                    "lastFour": tokenizedCard.lastFour,
                    "expirationMonth": tokenizedCard.expirationMonth,
                    "expirationYear": tokenizedCard.expirationYear,
                    "cardholderName": tokenizedCard.cardholderName
                ]
                self.sendSuccess(json: json, result: result)
            } else {
                result(FlutterError(code: Constants.ERROR_KEY, message: "Tokenize card failed", details: "No data and no error returned"))
            }
        }
    }
    
    private func sendSuccess(json: [String: Any?], result: @escaping FlutterResult) {
        do {
            let data = try JSONSerialization.data(withJSONObject: json, options: [])
            let jsonString = String(data: data, encoding: .utf8)
            result(jsonString)
        } catch {
            result(FlutterError(code: Constants.ERROR_KEY, message: "Failed to serialize JSON", details: error.localizedDescription))
        }
    }
}

extension BraintreeCheckoutFlutterPlugin {
    public func application(
        _ application: UIApplication,
        continue userActivity: NSUserActivity,
        restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void
    ) -> Bool {
        guard userActivity.activityType == NSUserActivityTypeBrowsingWeb,
              let url = userActivity.webpageURL else {
            return false
        }

        if isBraintreeUniversalLink(url: url) {
            BTAppContextSwitcher.sharedInstance.handleOpen(url)
            return true
        }
        
        return false
    }
}

File: android/src/main/kotlin/io/navalia/braintree_checkout_flutter/BraintreeCheckoutFlutterPlugin.kt
package io.navalia.braintree_checkout_flutter

import android.app.Activity
import android.app.Activity.RESULT_OK
import android.content.Context
import android.content.Intent
import com.braintreepayments.api.card.Card
import com.braintreepayments.api.card.CardClient
import com.braintreepayments.api.card.CardResult
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import com.braintreepayments.api.datacollector.DataCollector
import com.braintreepayments.api.datacollector.DataCollectorRequest
import com.braintreepayments.api.core.BraintreeClient
import com.braintreepayments.api.core.DeviceInspector
import com.braintreepayments.api.paypal.PayPalResult
import org.json.JSONObject

class BraintreeCheckoutFlutterPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private var pendingResult: Result? = null
    private lateinit var context: Context

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "braintree_checkout_flutter")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            Constants.VENMO_PAYMENT_METHOD_KEY -> {
                val arguments = call.arguments as? Map<String, Any> ?: emptyMap()
                startActivity(
                    VenmoActivity::class.java,
                    Constants.VENMO_REQUEST_CODE,
                    arguments, result,
                )
            }

            Constants.PAYPAL_PAYMENT_METHOD_KEY -> {
                val arguments = call.arguments as? Map<String, Any> ?: emptyMap()
                startActivity(
                    PayPalActivity::class.java,
                    Constants.PAYPAL_REQUEST_CODE,
                    arguments, result,
                )
            }

            Constants.GET_DATA -> {
                val arguments = call.arguments as? Map<String, Any> ?: emptyMap()
                collectDeviceData(arguments, result)
            }

            Constants.IS_VENMO_APP_INSTALLED -> {
                val isVenmoAppInstalled = DeviceInspector().isVenmoAppSwitchAvailable(context)
                result.success(isVenmoAppInstalled)
            }

            Constants.TOKENIZE_CARD_METHOD_KEY -> {
                val arguments = call.arguments as? Map<String, Any> ?: emptyMap()
                tokenizeCard(arguments, result)
            }

            else -> result.notImplemented()
        }
    }

    private fun startActivity(
        activityClass: Class<out Activity>,
        activityRequestCode: Int,
        arguments: Map<String, Any>, result: Result
    ) {
        if (activity == null) {
            result.error(Constants.ERROR_KEY, "Activity is not available", null)
            return
        } 
        if (arguments.isEmpty()) {
            result.error(Constants.ERROR_KEY, "Arguments is not available", null)
            return
        }
        val intent = Intent(activity, activityClass)
        IntentUtils.putArguments(intent, arguments)
        pendingResult = result
        activity!!.startActivityForResult(intent, activityRequestCode)
    }

    fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
        if (requestCode == Constants.VENMO_REQUEST_CODE
            || requestCode == Constants.PAYPAL_REQUEST_CODE
        ) {
            when {
                resultCode == Activity.RESULT_OK && intent != null -> {
                    val nonce = intent.getStringExtra(Constants.NONCE_KEY)
                    pendingResult?.success(nonce)
                }
                resultCode == Activity.RESULT_CANCELED && intent != null -> {
                    val error = intent.getStringExtra(Constants.ERROR_KEY)
                    if (error != null) {
                        pendingResult?.error(Constants.ERROR_KEY, error, null)
                        return
                    }
                    val canceled = intent.getStringExtra(Constants.CANCELED_KEY)
                    if (canceled != null) {
                        pendingResult?.success(null)
                        return
                    }
                }
                else -> {
                    pendingResult?.success(null)
                }
            }
            pendingResult = null
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addActivityResultListener { requestCode, resultCode, data ->
            onActivityResult(requestCode, resultCode, data)
            true
        }
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onDetachedFromActivityForConfigChanges() {}

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    private fun collectDeviceData(arguments: Map<String, Any>, result: Result) {
        if (activity == null) {
            result.error(Constants.ERROR_KEY, "Activity is not available", null)
            return
        }
        try {
            val token = arguments[Constants.TOKEN_KEY] as? String
            if (token == null) {
                result.error(Constants.ERROR_KEY, "Token is required for data collection", null)
                return
            }
            val braintreeClient = BraintreeClient(activity!!, token)
            val dataCollector = DataCollector(braintreeClient)
            val dataCollectorRequest = DataCollectorRequest(hasUserLocationConsent = true)
            dataCollector.collectDeviceData(activity!!, dataCollectorRequest) { dataCollectorResult ->
                when (dataCollectorResult) {
                    is com.braintreepayments.api.datacollector.DataCollectorResult.Success -> {
                        result.success(dataCollectorResult.deviceData)
                    }
                    is com.braintreepayments.api.datacollector.DataCollectorResult.Failure -> {
                        result.error(Constants.ERROR_KEY, "Device data collection failed", dataCollectorResult.error.localizedMessage)
                    }
                }
            }
        } catch (e: Exception) {
            result.error(Constants.ERROR_KEY, "Error during device data collection", e.localizedMessage)
        }
    }

    private fun tokenizeCard(arguments: Map<String, Any>, result: Result) {
        if (activity == null) {
            result.error(Constants.ERROR_KEY, "Activity is not available", null)
            return
        }
        try {
            val token = arguments[Constants.TOKEN_KEY] as? String
            val cardholderName = arguments[Constants.CARDHOLDER_NAME_KEY] as? String
            val cardNumber = arguments[Constants.CARD_NUMBER_KEY] as? String
            val expirationMonth = arguments[Constants.EXPIRATION_MONTH_KEY] as? String
            val expirationYear = arguments[Constants.EXPIRATION_YEAR_KEY] as? String
            val cvv = arguments[Constants.CVV_KEY] as? String
            if (token == null || cardNumber == null || expirationMonth == null || expirationYear == null || cvv == null) {
                result.error(Constants.ERROR_KEY, "Token, card number, expiration month, expiration year and cvv are required for tokenization", null)
                return
            }
            val card = Card(
                cardholderName = cardholderName,
                number = cardNumber,
                expirationMonth = expirationMonth,
                expirationYear = expirationYear,
                cvv = cvv
            )
            val cardClient = CardClient(
                activity!!, token
            )
            cardClient.tokenize(card) { cardResult ->
                when (cardResult) {
                    is CardResult.Success -> {
                        result.success(parseSuccessResult(cardResult))
                    }

                    is CardResult.Failure -> {
                        result.error(Constants.ERROR_KEY, "Tokenize card failed", cardResult.error.localizedMessage)
                    }
                }
            }
        } catch (e: Exception) {
            result.error(Constants.ERROR_KEY, "Error during device data collection", e.localizedMessage)
        }
    }

     private fun parseSuccessResult(result: CardResult.Success) : String {
        val nonce = result.nonce
        val nonceJson = JSONObject().apply {
            put("nonce", nonce.string)
            put("isDefault", nonce.isDefault)
            put("cardType", nonce.cardType)
            put("lastTwo", nonce.lastTwo)
            put("lastFour", nonce.lastFour)
            put("expirationMonth", nonce.expirationMonth)
            put("expirationYear", nonce.expirationYear)
            put("cardholderName", nonce.cardholderName)
        }
        return nonceJson.toString()
    }
}
